src/frequenz/sdk/_data_ingestion/battery_utils.py

```diff
@@ -0,0 +1,334 @@
+        """Async part of constructor.
+
+        This should be called in order to subscribe for necessary data.
+
+        Raises:
+            RuntimeError: If given battery as no adjacent inverter.
+        """
+        component_graph = get_microgrid().component_graph
+        inverter_id = self._find_adjacent_inverter_id(component_graph)
+        if inverter_id is None:
+            raise RuntimeError(
+                f"Can't find inverter adjacent to battery: {self._battery_id}"
+            )
+
+        api_client = get_microgrid().api_client
+
+        bat_recv = await api_client.battery_data(self._battery_id)
+        # Defining battery_receiver and inverter receiver needs await.
+        # Because of that it is impossible to define it in constructor.
+        # Setting it to None first would require to check None condition in
+        # every call.
+        # pylint: disable=attribute-defined-outside-init
+        self._battery_receiver = _ComponentReceiver(
+            self._battery_id, bat_recv.into_peekable()
+        )
+        inv_recv = await api_client.inverter_data(inverter_id)
+        # pylint: disable=attribute-defined-outside-init
+        self._inverter_receiver = _ComponentReceiver(
+            inverter_id, inv_recv.into_peekable()
+        )
+
+        await asyncio.sleep(_START_DELAY_SEC)
+        self._init_method_called = True
+
+    def is_ready(self) -> bool:
+        """Return if battery message indicates that battery can be used.
+
+        Raises:
+            RuntimeError: if method `init` was not called or not awaited.
+
+        Returns:
+            True if battery can be used, False otherwise.
+        """
+        if not self._init_method_called:
+            raise RuntimeError(
+                "`init` method not called or not awaited. Run it before first use"
+            )
+
+        bat_msg = self._battery_receiver.receiver.peek()
+        inv_msg = self._inverter_receiver.receiver.peek()
+        if bat_msg is None or inv_msg is None:
+            component_id = self._battery_id
+            if inv_msg is None:
+                component_id = self._inverter_receiver.component_id
+            _logger.warning("None returned from component %d receiver.", component_id)
+            return False
+
+        older_message = bat_msg if bat_msg.timestamp < inv_msg.timestamp else inv_msg
+        is_ready = (
+            self._is_message_reliable(older_message)
+            and self._is_battery_state_correct(bat_msg)
+            and self._is_inverter_state_correct(inv_msg)
+            and self._no_critical_error(bat_msg)
+            and self._no_critical_error(inv_msg)
+        )
+        # If message is correct, but it wasn't before, then unblock it.
+        # This will allow to use battery shortly after it start working.
+        if is_ready and not self._last_ready:
+            self.unblock()
+        self._last_ready = is_ready
+
+        return is_ready
+
+    def is_blocked(self) -> bool:
```

Comment:
From an UX perspective, it would seem that `!battery.is_ready()` would achieve the same purpose. But it would not, therefore introducing a bit of inconsistency.

Looks like the timestamp check is not a part of the `is_ready` method. If you add it there, then the `is_blocked` method can be removed.

Reply:
But then I can't get batteries that are blocked and ready.

Reply:
I can't just move it, because then I won't be able to get batteries with correct message but are blocked.

I renamed `is_ready` to `get_state` that returns ENUM with information if battery is working, blocked or not working.